main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -18,264 +18,564 @@ package ca.uwaterloo.flix.language.phase
+  * A function is an entry point if:
+  *   - It is the main function (called `main` by default, but can configured to an arbitrary name).
+  *   - It is a test (annotated with `@Test`).
+  *   - It is an exported function (annotated with `@Export`).
   *
-  * The argument to the `func` must have type `Unit`.
-  *
-  * If the result type of `func` has type `Unit`,
-  * then the result is returned from `main%` as normal.
-  * If the result type of `func` is some other type with a `ToString` instance,
-  * then the result is printed in `main%`.
-  * If the result type of `func` is some other type without a `ToString` instance,
-  * then an error is raised.
-  *
-  * For example, given an entry point `func` with type `Unit -> Float64`,
-  * we produce:
-  * {{{
-  *  pub def main%(): Unit \ IO = {
-  *      println(func(args))
-  *  }
-  * }}}
+  * This phase has four sub-phases:
+  *   1. Resolve the entrypoint option such that there is no implicit default entry point.
+  *   1. Check that all entry points have valid signatures, where rules differ from main, tests, and
+  *      exports. If an entrypoint does not have a valid signature, its related annotation is
+  *      removed to allow further compilation to continue with valid assumptions.
+  *   1. Replace the existing main function by a new main function that prints the returned value if
+  *      its return type is not Unit.
+  *   1. Compute the set of all entry points and store it in Root.
   */
 object EntryPoint {
 
   // We don't use regions, so we are safe to use the global scope everywhere in this phase.
   private implicit val S: Scope = Scope.Top
 
-  /**
-    * The scheme of the entry point function.
-    * `Unit -> Unit`
-    */
-  private val EntryPointScheme = Scheme(Nil, Nil, Nil, Type.mkIoArrow(Type.Unit, Type.Unit, SourceLocation.Unknown))
-
-  /**
-    * The default entry point in case none is specified. (`main`)
-    */
-  private val DefaultEntryPoint = Symbol.mkDefnSym("main")
-
-  /**
-    * Introduces a new function `main%` which calls the entry point (if any).
-    */
   def run(root: TypedAst.Root)(implicit flix: Flix): (TypedAst.Root, List[EntryPointError]) = flix.phaseNew("EntryPoint") {
-    implicit val sctx: SharedContext = SharedContext.mk()
-    val newRoot = findOriginalEntryPoint(root) match {
-      // Case 1: We have an entry point. Wrap it.
-      case Some(entryPoint0) =>
-        val entryPoint = visitEntryPoint(entryPoint0, root, TraitEnv(root.traitEnv))
-        root.copy(
-          defs = root.defs + (entryPoint.sym -> entryPoint),
-          entryPoint = Some(entryPoint.sym),
-          reachable = getReachable(root) + entryPoint.sym
-        )
-      // Case 2: No entry point. Don't touch anything.
-      case None => root.copy(reachable = getReachable(root))
-    }
-    (newRoot, sctx.errors.asScala.toList)
+    val (root1, errs1) = resolveMain(root)
```

Comment:
I don't hate that you have these subphases, but I do hate that it's not consistent what's a subphase and what's just a function.

Reply:
Do we really need to think of phases? Whats wrong with just 4 functions?

Reply:
And why do they have to be in objects? We stopped doing that. Flat structure with functions please :)

Reply:
I think it makes the phase readable, otherwise its just a soup of functions. Now you know which functions belong together

4 phases is a good thing because now each "phase" has documented rules and changes - makes the whole phase easier to have in your head